docs/design/storagenode-graceful-exit.md

```diff
@@ -91,6 +92,133 @@ This process including the Storage nodes transferring their pieces to other node
+
+#### Storagenode
+- Add `gracefulexit` command to storagenode CLI
+  - When executed, the user should be prompted with a numbered list of satellites (what would we display as a name?)
+  - After selecting a satellite, the user should be prompted for a confirmation
+  - Once confirmed, the command should call the `GracefulExit.Initiate` endpoint for that satellite
+  - Records starting disk usage in `gexit.exit_status`
+- Add `gexit` DB implementation with `exit_order`  and `exit_order` tables
+  - ```
+	model exit_order (
+		key satellite_id serial_number
+
+		field satellite_id            blob not null
+		field serial_number           blob not null
+		field order_limit_serialized  blob not null
+		filed order_limit_expiration  timestamp not null
+		field completed_dt            timestamp ( updateable )
+	)
+
+	model exit_status (
+		key satellite_id
+
+		field satellite_id           blob not null
+		field initiated_dt           timestamp ( autoinsert ) not null
+		field completed_dt           timestamp ( updateable )
+		field starting_disk_usage    int64 not null
+		field bytes_deleted          int64
+	)	
+	```
+- Add GracefulExit endpoint
+  - Endpoint used by satellites to tell the exiting storage node what pieces to process
+
+  - ``` 
+	service GracefulExit {
+		// Called by the satellite to batch piece orders to be moved to new nodes
+		rpc ProcessOrders(OrderRequest) returns (OrderResponse) {}
+		// Called by the satellite to notify the storagenode that the exit is complete for this satellite
+		rpc Completed(CompletedRequest) returns (CompletedResponse)
+		// Called by the satellite to get exit status information
+		rpc Status(StatusRequest) returns (StatusResponse) {}
+	}
+
+	message Order {
+		bytes hashing_key
+		AddressedOrderLimit addressed_order_limit 
+	}
+
+	message OrderRequest {
+		Order orders repeatable
+	}
+
+	message OrderResponse {
+	}
+
+    message CompletedRequest {
+		google.protobuf.Timestamp completed_dt
+	}
+
+    message CompletedResponse {
+	}
+
+	message StatusRequest {
+	}
+
+	message StatusResponse {
+        byte satellite_id
+        google.protobuf.Timestamp initiated_dt
+        google.protobuf.Timestamp completed_dt
+        int64 starting_disk_usage
+        int64 bytes_deleted		
+	}
+	```
+- Update bandwidth usage monitors to ignore `PieceAction_PUT_EXIT` bandwidth actions
+- Add GracefulExit service
+  - Iterates over `exit_order` where `completed_dt` is null
+    - Pushes the pieces to the storage node identified in the order using `ecclient`
+    - Sends the signed new node response to the satellite via a new `CommitPiece` method (uses `metainfo.UpdatePieces`).
+    - Updates `bytes_deleted` with the number of bytes deleted and sets `completed_dt` to current time
+  - Execution intervals and batch sizes should be configurable
 
 ## Open Questions
```

Comment:
"What happens if a piece is deleted when the piece is in the process of being transferred from the exiting node to other node in the network?"

I believe if I understood this correctly, the exiting storage node shall only honor download request and not delete requests or new upload request, Infact exiting SN should internal be made to accept only the download and reject accepting new uploads and/or delete requests